### ISSUE TITLE
produce rate is decreasing a lot when leader is changed/restarted

### ISSUE THREAD

**Issue by silviucpp** [kafka4beam/brod#155]
Hello,

We observed in production when we restarted the leader some strange behavior where the produced message rate decreased a lot and all caller processes for produce method were blocked in that call.

Also I did a small script to send messages and after we restart the leader or try to move partition on another machine we can see the same behavior. It's taking over 30 seconds to get everything back to normal.

https://dl.dropboxusercontent.com/u/30548263/test_kafka.erl

     test_kafka:init().
     test_kafka:run_infinite(8, 10000, 0).

This will send in loop using 8 erlang processes messages to kafka on topic 0. If you look in observer (will be opened as well) after you restart the topic the messages queue for the brod processes are staying freezed and are cleaned after 30 -40 seconds. Also if you look in kafka the number of messages produced is decreasing a lot.

Any technical reason for this ? or any missed configure on our side ?

Silviu

**Comment by zmstone** [comment]
Hi @silviucpp 

A part of this is a design intention (right or wrong is open for discussion).
Up on a failure, brod_producer will re-queue the failed messages and start over.
The retry is implemented in a 'slow' fashion: It sends sync-requests for one message at a time.
For two reason
1. To start slow (Looking at it retrospectively now, It's probably not necessary, will send a PR to fix it)
2. For message strict ordering

BTW. Leader re-discovery might have contributed a bit to your 30-40 seconds?

As a work around before we can fix it with a better batching, you can:
1. Batch on the caller side: `call brod:prodcue(Client, Topic, Patition, _Key = <<>>, _Value = [{Key, Value}])`:  so brod_producer can actually retry a batch (instead of one message) at a time.
2. Fail fast if the caller has the ability to retry (or if it is ok to have few messages getting lost): brod_producer by default has 3 reties, if `max_retries` given in producer config is 0, brod_producer will crash immediately, then supervisor should restart it (after a configurable sleep).

**Comment by silviucpp** [comment]
Hello,

Leader discovery I don't think has any contribution as time I can see it's discovered almost instantly. 

Do you have any idea when an improvement in this are is planned ? In the meantime I will try your second suggestion

Silviu

**Comment by zmstone** [comment]
I will send the pr in a day or two.

**Comment by silviucpp** [comment]
I just tested with {max_retries, 0} as well now but I still see that poor behavior (low rate, caller blocked for a ling time). 
Instantly when we restart the leader I can see: client client1 connected to 10.11.5.44:9092 which I suppose the leader is discovered instantly (10.11.5.44 being the new leader).

Silviu

**Comment by zmstone** [comment]
Just to be sure: you should see `reached_max_retries` crash logs.

`partition_restart_delay_seconds` is to configure how long producer supervisor should wait before restarting the (per partition) worker (by default it's 5 seconds, please try 1 see if it makes any difference).

**Comment by silviucpp** [comment]
hmm I don't see this.. 

     ok = brod:start_producer(client1, ?TOPIC, [
           {reconnect_cool_down_seconds, 10},
           {default_producer_config, [{required_acks, 0}, {max_retries, 0}]}
      ]).

This is what I've used.. I'm doing something wrong ?
Silviu

**Comment by zmstone** [comment]
try add `{partition_restart_delay_seconds, 1}` to your `default_producer_config`

**Comment by silviucpp** [comment]
Ok So I used:

```erlang
    ok = brod:start_producer(client1, ?TOPIC, [
        {reconnect_cool_down_seconds, 10},
        {default_producer_config, [
            {required_acks, 0},
            {max_retries, 0},
            {partition_restart_delay_seconds, 1}
        ]}
    ]).
```

And I only see:

```
=INFO REPORT==== 30-Jan-2017::15:42:55 ===
client client1 connected to 10.11.5.45:9092

=ERROR REPORT==== 30-Jan-2017::15:43:31 ===
Error in produce response
Topic: topic-test
Partition: 0
Offset: 1338761
Error: 'NotLeaderForPartition'3> 
=INFO REPORT==== 30-Jan-2017::15:43:31 ===
client client1: payload socket down 10.11.5.45:9092
reason:{shutdown,tcp_closed}3> 
=INFO REPORT==== 30-Jan-2017::15:43:32 ===
client client1 connected to 10.11.5.44:9092
```

**Comment by zmstone** [comment]
ah
```eralng
start_producer(client1, ?TOPIC,
           [{required_acks, 0},
            {max_retries, 0},
            {partition_restart_delay_seconds, 1}])
```
reconnect_cool_down_seconds
default_producer_config
they are client configs.

**Comment by silviucpp** [comment]
Hello,

Thanks a lot,

Seems also `required_acks` was in the wrong place. If I keep it 0 everything is fine, I see in logs:

unexpected response for partition: 0 error: {error,{producer_down,noproc}} 

So that issue takes place only in case require_ack is not 0.

Silviu

**Comment by zmstone** [comment]
right, required acks is a client config.
and yes, when required acks is 0, the retry (if ever will happen) should be very fast (as it doesn't wait for kafka to ack).

Please keep this issue open for though.
I'll close it when the fix is merged.

**Comment by zmstone** [comment]
sorry, bad memory, required_acks is a producer config.

**Comment by silviucpp** [comment]
I'm confused.. I don't understand why require_acks is not working when I'm using app.config
```erlang
{brod,
        [{clients, [
            {client1, [
                {endpoints, [{"10.11.5.41", 9092}]},
                {auto_start_producers, true},
                {config, [
                    %{reconnect_cool_down_seconds, 10},
                    {default_producer_config, [
                        {required_acks, 0}
                    ]}
                ]}
            ]}
        ]}]
    },
```

But working when I set config from code:

```erlang
    ok = brod:start_client(?ENDPOINTS, client1),
    ok = brod:start_producer(client1, ?TOPIC, [
        {required_acks, 0}
    ]).
```

Any idea ?

**Comment by silviucpp** [comment]
I tried it in config directly as well:

```erlang
{brod,
        [{clients, [
            {client1, [
                {endpoints, [{"10.11.5.41", 9092}]},
                {auto_start_producers, true},
                {config, [
                    %{reconnect_cool_down_seconds, 10},
                    {required_acks, 0}
                ]}
            ]}
        ]}]
    },
```

And still not working

**Comment by zmstone** [comment]
```erlang

            {client1, [
                {endpoints, [{"10.11.5.41", 9092}]},
                {auto_start_producers, true},

                    %{reconnect_cool_down_seconds, 10},
                    {required_acks, 0}
           
            ]}

```

**Comment by silviucpp** [comment]
Thanks, 

I already figure out looking to the code. But your `sys.config.example` from git seems outdated..

Silviu

**Comment by zmstone** [comment]
damn! I was so fixated on README.md mised sys.config.example.
will fix that.
sorry for the confusion, and thanks for reporting it.

**Comment by silviucpp** [comment]
Hmm... I might found another bug..
I created a small load test app to see how many messages /second I can produce.. 
when using require_ack 0 it's around 6k/second.. when using require_ack 1 it's 67k/second..

Any idea why is so slow with 0 ? I was expected to be damn fast ..

**Comment by zmstone** [comment]
it might related to batching.
when required acks is 0, it might not batch at all.
we have #104 not fixed yet.

**Comment by silviucpp** [comment]
Most probably :) I'll test again after #104 is fixed

**Comment by silviucpp** [comment]
Hello,

Do you have any load test script ? I wrote one but whatever I do I can't produce more than 70K messages/second with brod.

We tested also java driver and on the same cluster we got around 700 K/s and also the c driver and we got 1M/s .

We did all testing with require_ack 1 on one single partition . All other settings being defaults.
Also we sent same message.

The erlang code looks like:

```erlang
-module(test_kafka).

-define(TOPIC, <<"topic-test">>).
-define(ENDPOINTS_DEV, [{"172.17.3.107", 9092}]).
-define(ENDPOINTS, ?ENDPOINTS_DEV).

-export([init/0, run/3]).

loop(0, _) ->
    ok;
loop(ReqNr, Partition) ->
    case catch brod:produce(client1, ?TOPIC, Partition, <<"k1">>, <<"hello world!">>) of
        {ok, _} ->
            ok;
        Error ->
            io:format("unexpected response for partition: ~p error: ~p ~n", [Partition, Error])
    end,
    loop(ReqNr -1, Partition).


init() ->
    case application:ensure_all_started(brod) of
        {ok, [_H|_T] } ->
            ok = brod:start_client(?ENDPOINTS, client1),
            ok = brod:start_producer(client1, ?TOPIC, [
                {required_acks, 1}
            ]);
        _ ->
            {ok, []}
    end.

run(ClientsNr, ReqNr, Partition) ->
    init(),

    ReqPerProcess = round(ReqNr/ClientsNr),

    FunTest = fun()->
        loop(ReqPerProcess, Partition)
    end,

    {Time, _} = timer:tc(fun()-> do_work(FunTest, ClientsNr) end),

    TimeMs = Time/1000,
    TimeS = TimeMs/1000,
    io:format("## Test complete time: ~p ms => throughput ~p/s ~n", [TimeMs, ReqNr/TimeS]),
    ok.

do_work(Fun, Count) ->
    process_flag(trap_exit, true),
    spawn_childrens(Fun, Count),
    wait_responses(Count).

spawn_childrens(_Fun, 0) ->
    ok;
spawn_childrens(Fun, Count) ->
    spawn_link(Fun),
    spawn_childrens(Fun, Count -1).

wait_responses(0) ->
    ok;
wait_responses(Count) ->
    receive
        {'EXIT',_FromPid, _Reason} ->
            wait_responses(Count -1)
    end.
``` 

Something wrong with my test?

Silviu

**Comment by zmstone** [comment]
No, we do not have any load test script.
You can try to set larger `partition_onwire_limit` and `partition_buffer_limit` in producer config.
Max batch size is by default set to 1M, which is probably good enough.

`partition_onwire_limit` is the max number of async produce requests sent on wire (before receiving an ack) before brod can send another one.
It is set to 1 by default because we value strict ordering over throughput by default.

**Comment by silviucpp** [comment]
Hello,

I tried increasing them but no change... still bad performances.. 

I tried with partition_onwire_limit to 100000 and partition_buffer_limit to 32MB , also tried other different values or change them one by one.. 

Silviu

**Comment by zmstone** [comment]
I remember we tested with 1K payload, but didn't compare much with other libraries because the result was good enough for us.

You have probably pushed to the limit of the per-message overheads:
e.g. this is still a synced call to the producer buffer:
`brod:produce(client1, ?TOPIC, Partition, <<"k1">>, <<"hello world!">>)`

The caller can batch to have less roundtrips and reduce per-message overheads:
e.g.
```
Payload = lists:duplicate(10, {<<"k1">>, <<"hello world!">>}),
brod:produce(client1, ?TOPIC, Partition, <<>>, Payload)
```
I bet you'll see a much better result.

**Comment by id** [comment]
The bottleneck is indeed in synced call to brod_producer process. When I removed that sync point, I managed to reach throughput of 1M/sec from 40k/sec on my local machine.

Perhaps we can kill it and just send users 'ack' messages when we received a response from kafka? There is no much value in making sure a message reached another process.

**Comment by zmstone** [comment]
It more of a pushback to the callers.
I don't think it's a good idea to remove pushback (at least not by default) for below reasons:
1. Without pushback, when there is congestion on wire, the producer buffer or mailbox may explode.
2. There are alternatives to achieve better throughput, e.g. caller batch, more brod clients, more kafka partitions.
3. Messages of a few bytes of payload is not a practical use case.

Having all that said, however, if it is really needed, one can do it NOW by
first call `brod:get_producer/3` to get the pid,
then monitor the pid and send produce requests like this:

```
my_true_async_produce_api(Pid, Key, Value) ->
  CallRef = #brod_call_ref{ caller = self()
                          , callee = Pid
                          , ref    = make_ref()
                          },
  Pid ! {produce, CallRef, Key, Value},
  ok.
```
Then collect `#brod_produce_reply{}` asynchronously.

**Comment by silviucpp** [comment]
Hello,

We performed testing across all drivers we could (java one, scala one and the c driver librdkafka). From what I can tell is:

- the batch logic is not something that caller should handle . they handle everything internally, you can only tune the params. If you put this task on the caller side means each client using your library will have it's own implementation with it's own bugs. Nothing uniform and easy to use
- I tried to increase number of clients as well .. unfortunately there is no performance boost on brod.
- I tested also with multiple partitions and performances are not significant better
- I also tested with 1 KB messages and still there is at least 60 k Messages/s between brod and c driver

Feel free to change my load test program in case I did something wrong in calling brod API. I can rerun my tests after this.

Maybe instead using erlang mailbox to queue the requests you can use ets or something else. I think it's much faster as time form what I saw when  an erlang process has a lot of messages in it's inbox it's starting performing very slow.

Silviu

**Comment by zmstone** [comment]
Hi @silviucpp I have made some update based on your test code.
Please see #184 or branch named `load-test-2` for details.

I added a non-blocking version produce API `brod:produce_cast` to test how much it would improve.
The throughputs In my tests:
`test_kafka:run(1, 100000, 0, produce)` was around 80k.
`test_kafka:run(1, 100000, 0, produce_cast)` was around 110k.
`test_kafka:run(1, 100000, 0, pingpong)` was about 550k.
the `pingpong` result is like a baseline to see the cost of a dead-simple inter-process round-trip.

As for the ETS suggestion.
I'm not sure how that should work if we don't change the acking mechanism:
For each and every brod:produce call, there needs to be an ack.
The costs are really the number of message passing,
The size of the messages matters but it only gets worse if we need to copy them in/out ETS. 

Giving lager `partition_onwire_limit` results in less throughput,
I guess it's because it makes the batch sizes smaller, which in turn lowers the throughput.
I'll later add`min_batch_size` and `max_linger_age` producer configs.
then test with `test_kafka:run(1, OneMillion, 0, produce/produce_cast)` see if it'll make a difference.

**Comment by zmstone** [comment]
@id I'm not sure how you get the 1M throughput, if it's only about casting message without waiting for any reply? then it's point-less because no producer would actually work like that.

In my test code (based on @silviucpp's code), I try to collect all the acks from kafka before exiting the loop.

**Comment by zmstone** [comment]
@silviucpp This issue has been (auto) closed.
Please feel free to open a new issue if you think we should continue discussing throughput.
